test/should_fail/map_field_invalid_update.erl

```diff
@@ -0,0 +1,10 @@
+-module(map_field_invalid_update).
+
+-export([f/1]).
+
+-type map_with_integer() :: #{a := integer()}.
+-type map_with_binary()  :: #{a := binary()}.
+
+-spec f(map_with_integer()) -> map_with_binary().
```

Comment:
You could write `#{a := integer()}` and `#{a := binary()}` directly in the spec, to keep user-defined types out of the test case. Just to make sure the user-def types handling is not affecting anything.